### PR TITLE
Write Promtail position file atomically.

### LINKED ISSUE #5218: Promtail: recover from corrupted position file
If Promtail crash while saving the positions file, this one can end up corrupted.

I suggested we always write first to new temporary file `position.yaml.new` and then `os.Rename` which I think is atomic on most platform.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
This uses `renameio` that supports writing files atomically.

**Which issue(s) this PR fixes**:
Fixes #5218

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [x] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: clients/pkg/promtail/positions/positions.go [hunk 1]
```diff
@@ -252,12 +253,11 @@ func writePositionFile(filename string, positions map[string]string) error {
 	}
 
 	target := filepath.Clean(filename)
-	temp := target + "-new"
 
-	err = ioutil.WriteFile(temp, buf, os.FileMode(positionFileMode))
+	err = renameio.WriteFile(target, buf, os.FileMode(positionFileMode))
```
Comment (dannykopping) [inline clients/pkg/promtail/positions/positions.go:257]: https://github.com/google/renameio#windows-support

According to this, we should probably special-case Windows here. What do you think?

**Review by dannykopping — COMMENTED** [review]
Nice! Added a question about Windows but overall LGTM

File: go.mod [hunk 2]
```diff
@@ -179,6 +180,7 @@ require (
 	github.com/google/go-querystring v1.0.0 // indirect
 	github.com/google/gofuzz v1.1.0 // indirect
 	github.com/google/pprof v0.0.0-20211214055906-6f57359322fd // indirect
+	github.com/google/renameio v0.1.0 // indirect
```
Comment (jeschkies) [inline go.mod:183]: I wonder where v0 is coming from...

File: clients/pkg/promtail/positions/positions.go (hunk 1 above)
Comment (jeschkies) [reply clients/pkg/promtail/positions/positions.go:257]: I wanted to reference it. You were faster. I think so. We should use the old version on Windows.

File: clients/pkg/promtail/positions/write_positions.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	}
+
+	target := filepath.Clean(filename)
+
+	err = renameio.WriteFile(target, buf, os.FileMode(positionFileMode))
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (chaudum) [inline clients/pkg/promtail/positions/write_positions.go:26]: You could simply return the return value from `renameio.WriteFile()`

```suggestion
	return renameio.WriteFile(target, buf, os.FileMode(positionFileMode))
```

**Review by chaudum — APPROVED** [review]
Minor suggestion, but otherwise LGTM!

File: go.mod (hunk 2 above)
Comment (dannykopping) [reply go.mod:183]: What happens when you run `go mod tidy`?

File: clients/pkg/promtail/positions/write_positions_windows.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+package positions
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	yaml "gopkg.in/yaml.v2"
+)
+
+func writePositionFile(filename string, positions map[string]string) error {
```
Comment (dannykopping) [inline clients/pkg/promtail/positions/write_positions_windows.go:11]: An explanation of why this function is duplicated for Windows would be good for future reference.

File: go.mod (hunk 2 above)
Comment (jeschkies) [reply go.mod]: Did the trick!

File: clients/pkg/promtail/positions/write_positions.go (hunk 3 above)
Comment (jeschkies) [reply clients/pkg/promtail/positions/write_positions.go:26]: :see_no_evil:

**Pull request merged**